### ISSUE TITLE
truncation in conversion from uint64_t to size_t in StringRef::substr()

### ISSUE THREAD

**Issue by msebor** [llvm/llvm-project#54739]
As pointed out in https://reviews.llvm.org/D122836#3423517, passing the `uint64_t` result of `getZExtValue()` to `StringRef::substr(size_t, size_t)` as is done below might silently truncate the value in ILP32.  This is just a reminder to review and fix that and any other calls like it.

```
lib/Transforms/Utils/SimplifyLibCalls.cpp:  Str = Str.substr(0, LenC->getZExtValue());
```

It looks like the code in `DWARFUnit::tryExtractDIEsIfNeeded` in `llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp` might be subject to the sametruncation:
```
    if (auto *IndexEntry = Header.getIndexEntry())
      if (const auto *C = IndexEntry->getContribution(
              Header.getVersion() >= 5 ? DW_SECT_LOCLISTS : DW_SECT_EXT_LOC))
        Data = Data.substr(C->Offset, C->Length);
```

**Comment by msebor** [comment]
This is starting to look like fairly wide-spread problem.  Rebuilding `opt` and `clang` with the following patch:
```
diff --git a/llvm/include/llvm/ADT/StringRef.h b/llvm/include/llvm/ADT/StringRef.h
index 118def2f43e1..deccbc1cdfe3 100644
--- a/llvm/include/llvm/ADT/StringRef.h
+++ b/llvm/include/llvm/ADT/StringRef.h
@@ -613,6 +613,16 @@ namespace llvm {
       return StringRef(Data + Start, std::min(N, Length - Start));
     }
 
+    template <class WideInt1, class WideInt2>
+    __attribute__ ((warning ("silent truncation")))
+    StringRef substr(WideInt1 Start, WideInt2 N = npos,
+                    std::enable_if_t<(sizeof (WideInt1) > sizeof (size_t) ||
+                                      sizeof (WideInt2) > sizeof (size_t)), bool> = true) const
+    {
+      Start = std::min((size_t)Start, (size_t)Length);
+      return StringRef(Data + Start, std::min((size_t)N, Length - (size_t)Start));
+    }
+
     /// Return a StringRef equal to 'this' but with only the first \p N
     /// elements remaining.  If \p N is greater than the length of the
     /// string, the entire string is returned.
```
I get the warnings below.  I have only reviewed a few but I'm guessing they may not all be indicative of bugs, and some of those that are bugs might be latent, but their number makes me think that rather than trying to fix the call sites a better approach might be to change the `StringRef` API to take `uint64_t` instead of `size_t`.
```
/src/llvm/llvm/lib/Support/DataExtractor.cpp: In member function ‘llvm::StringRef llvm::DataExtractor::getBytes(uint64_t*, uint64_t, llvm::Error*) const’:
/src/llvm/llvm/lib/Support/DataExtractor.cpp:189:33: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  189 |   StringRef Result = Data.substr(*OffsetPtr, Length);
--
/src/llvm/llvm/lib/TableGen/Record.cpp: In member function ‘llvm::Init* llvm::TernOpInit::Fold(llvm::Record*) const’:
/src/llvm/llvm/lib/TableGen/Record.cpp:1423:29: warning: call to ‘llvm::StringRef::substr<long long int, long long int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
 1423 |       return StringInit::get(LHSs->getValue().substr(Start, Length),
--
/src/llvm/llvm/lib/DWARFLinker/DWARFLinker.cpp: In member function ‘llvm::DIE* llvm::DWARFLinker::DIECloner::cloneDIE(const llvm::DWARFDie&, const llvm::DWARFFile&, llvm::CompileUnit&, llvm::OffsetsStringPool&, int64_t, uint32_t, unsigned int, bool, llvm::DIE*)’:
/src/llvm/llvm/lib/DWARFLinker/DWARFLinker.cpp:1383:77: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
 1383 |   SmallString<40> DIECopy(Data.getData().substr(Offset, NextOffset - Offset));
--
/src/llvm/llvm/lib/DWARFLinker/DWARFLinker.cpp: In member function ‘void llvm::DWARFLinker::patchFrameInfoForObject(const llvm::DWARFFile&, llvm::RangesTy&, llvm::DWARFContext&, unsigned int)’:
/src/llvm/llvm/lib/DWARFLinker/DWARFLinker.cpp:1879:43: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
 1879 |       StringRef CIEData = FrameData.substr(EntryOffset, InitialLength + 4);
      |                           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/src/llvm/llvm/lib/DWARFLinker/DWARFLinker.cpp:1923:29: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
 1923 |     TheDwarfEmitter->emitFDE(IteratorInserted.first->getValue(), AddrSize,
--
/src/llvm/llvm/lib/DWARFLinker/DWARFStreamer.cpp: In member function ‘virtual void llvm::DwarfStreamer::emitLocationsForUnit(const llvm::CompileUnit&, llvm::DWARFContext&, std::function<void(llvm::StringRef, llvm::SmallVectorImpl<unsigned char>&)>)’:
/src/llvm/llvm/lib/DWARFLinker/DWARFStreamer.cpp:495:45: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  495 |       StringRef Input = InputSec.Data.substr(Offset, Length);
--
/src/llvm/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp: In member function ‘llvm::Value* llvm::LibCallSimplifier::optimizeStrNCmp(llvm::CallInst*, llvm::IRBuilderBase&)’:
/src/llvm/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:454:36: warning: call to ‘llvm::StringRef::substr<int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  454 |     StringRef SubStr1 = Str1.substr(0, Length);
      |                         ~~~~~~~~~~~^~~~~~~~~~~
/src/llvm/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:455:36: warning: call to ‘llvm::StringRef::substr<int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  455 |     StringRef SubStr2 = Str2.substr(0, Length);
--
/src/llvm/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp: In member function ‘llvm::Value* llvm::LibCallSimplifier::optimizeMemChr(llvm::CallInst*, llvm::IRBuilderBase&)’:
/src/llvm/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:935:19: warning: call to ‘llvm::StringRef::substr<int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  935 |   Str = Str.substr(0, LenC->getZExtValue());
--
/src/llvm/llvm/lib/Object/MachOUniversal.cpp: In member function ‘llvm::Expected<std::unique_ptr<llvm::object::MachOObjectFile> > llvm::object::MachOUniversalBinary::ObjectForArch::getAsObjectFile() const’:
/src/llvm/llvm/lib/Object/MachOUniversal.cpp:76:35: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
   76 |     ObjectData = ParentData.substr(Header64.offset, Header64.size);
--
/src/llvm/llvm/lib/Object/MachOUniversal.cpp: In member function ‘llvm::Expected<std::unique_ptr<llvm::object::IRObjectFile> > llvm::object::MachOUniversalBinary::ObjectForArch::getAsIRObject(llvm::LLVMContext&) const’:
/src/llvm/llvm/lib/Object/MachOUniversal.cpp:95:35: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
   95 |     ObjectData = ParentData.substr(Header64.offset, Header64.size);
--
/src/llvm/llvm/lib/Object/MachOUniversal.cpp: In member function ‘llvm::Expected<std::unique_ptr<llvm::object::Archive> > llvm::object::MachOUniversalBinary::ObjectForArch::getAsArchive() const’:
/src/llvm/llvm/lib/Object/MachOUniversal.cpp:114:35: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  114 |     ObjectData = ParentData.substr(Header64.offset, Header64.size);
--
/src/llvm/llvm/lib/Object/MachOObjectFile.cpp: In member function ‘llvm::ArrayRef<unsigned char> llvm::object::MachOObjectFile::getSectionContents(uint32_t, uint64_t) const’:
/src/llvm/llvm/lib/Object/MachOObjectFile.cpp:1966:31: warning: call to ‘llvm::StringRef::substr<unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
 1966 |   return arrayRefFromStringRef(getData().substr(Offset, Size));
--
/src/llvm/llvm/lib/DebugInfo/GSYM/FunctionInfo.cpp: In static member function ‘static llvm::Expected<llvm::gsym::FunctionInfo> llvm::gsym::FunctionInfo::decode(llvm::DataExtractor&, uint64_t)’:
/src/llvm/llvm/lib/DebugInfo/GSYM/FunctionInfo.cpp:69:49: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
   69 |                            Data.getAddressSize());
--
/src/llvm/llvm/lib/DebugInfo/GSYM/FunctionInfo.cpp: In static member function ‘static llvm::Expected<llvm::gsym::LookupResult> llvm::gsym::FunctionInfo::lookup(llvm::DataExtractor&, const llvm::gsym::GsymReader&, uint64_t, uint64_t)’:
/src/llvm/llvm/lib/DebugInfo/GSYM/FunctionInfo.cpp:191:54: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  191 |     const StringRef InfoBytes = Data.getData().substr(Offset, InfoLength);
--
/src/llvm/llvm/lib/DWP/DWP.cpp: In function ‘void addAllTypesFromTypesSection(llvm::MCStreamer&, llvm::MapVector<long long unsigned int, llvm::UnitIndexEntry>&, llvm::MCSection*, const std::vector<llvm::StringRef>&, const llvm::UnitIndexEntry&, uint32_t&)’:
/src/llvm/llvm/lib/DWP/DWP.cpp:242:20: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  242 |       Out.emitBytes(Types.substr(PrevOffset, C.Length));
--
/src/llvm/llvm/lib/DWP/DWP.cpp: In function ‘llvm::Error llvm::write(llvm::MCStreamer&, llvm::ArrayRef<std::__cxx11::basic_string<char> >)’:
/src/llvm/llvm/lib/DWP/DWP.cpp:642:78: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  642 |               parseInfoSectionUnitHeader(Info.substr(UnitOffset, Info.size()));
      |                                                                              ^
/src/llvm/llvm/lib/DWP/DWP.cpp:658:68: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  658 |                                  CurStrOffsetSection, CurStrSection);
      |                                                                    ^
/src/llvm/llvm/lib/DWP/DWP.cpp:676:24: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  676 |           Out.emitBytes(Info.substr(UnitOffset - C.Length, C.Length));
--
/src/llvm/llvm/lib/ProfileData/SampleProfReader.cpp: In function ‘bool ParseLine(const llvm::StringRef&, LineType&, uint32_t&, uint64_t&, uint32_t&, uint32_t&, llvm::StringRef&, llvm::DenseMap<llvm::StringRef, long long unsigned int>&, uint64_t&, uint32_t&)’:
/src/llvm/llvm/lib/ProfileData/SampleProfReader.cpp:208:47: warning: call to ‘llvm::StringRef::substr<unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  208 |         StringRef WordAfterColon = Rest.substr(n3 + 1, n4 - n3 - 1);
--
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp: In member function ‘llvm::Error llvm::coverage::RawCoverageReader::readString(llvm::StringRef&)’:
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:96:23: warning: call to ‘llvm::StringRef::substr<int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
   96 |   Result = Data.substr(0, Length);
--
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp: In member function ‘llvm::Error llvm::coverage::RawCoverageFilenamesReader::read(llvm::coverage::CovMapVersion)’:
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:130:48: warning: call to ‘llvm::StringRef::substr<int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  130 |     StringRef CompressedFilenames = Data.substr(0, CompressedLen);
--
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp: In member function ‘llvm::StringRef llvm::InstrProfSymtab::getFuncName(uint64_t, size_t)’:
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:469:21: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  469 |   return Data.substr(Pointer - Address, Size);
--
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp: In function ‘llvm::Expected<std::unique_ptr<llvm::coverage::BinaryCoverageReader> > loadTestingFormat(llvm::StringRef, llvm::StringRef)’:
/src/llvm/llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:885:78: warning: call to ‘llvm::StringRef::substr<int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  885 |   if (Error E = ProfileNames.create(Data.substr(0, ProfileNamesSize), Address))
--
/src/llvm/llvm/tools/dsymutil/BinaryHolder.cpp: In function ‘std::vector<llvm::MemoryBufferRef> llvm::dsymutil::getMachOFatMemoryBuffers(llvm::StringRef, llvm::MemoryBuffer&, llvm::object::MachOUniversalBinary&)’:
/src/llvm/llvm/tools/dsymutil/BinaryHolder.cpp:38:39: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
   38 |     StringRef ObjData = FatData.substr(It->getOffset(), It->getSize());
--
/src/llvm/llvm/tools/llvm-cxxdump/llvm-cxxdump.cpp: In function ‘void dumpCXXData(const llvm::object::ObjectFile*)’:
/src/llvm/llvm/tools/llvm-cxxdump/llvm-cxxdump.cpp:215:47: warning: call to ‘llvm::StringRef::substr<long long unsigned int, long long unsigned int>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  215 |     StringRef SymContents = SecContents.substr(SymOffset, SymSize);
      |                             ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/src/llvm/llvm/tools/llvm-cxxdump/llvm-cxxdump.cpp:330:31: warning: call to ‘llvm::StringRef::substr<long long unsigned int, unsigned char>’ declared with attribute warning: silent truncation [-Wattribute-warning]
  330 |             SymContents.substr(SymOffI, BytesInAddress).data();
```

**Comment by msebor** [comment]
I feel like I've opened up a can of worms.  The truncation is pervasive in the uses of many `StringRef` members.  The class uses `size_t` for string offsets and sizes (both to represent slices of strings and to  manipulate them) but many of its clients use `uint64_t`.  For instance here:
```
StringRef DataExtractor::getCStrRef(uint64_t *OffsetPtr, Error *Err) const {
  ErrorAsOutParameter ErrAsOut(Err);
  if (isError(Err))
    return StringRef();

  uint64_t Start = *OffsetPtr;
  StringRef::size_type Pos = Data.find('\0', Start);      <<< Start truncated
  if (Pos != StringRef::npos) {
    *OffsetPtr = Pos + 1;
    return StringRef(Data.data() + Start, Pos - Start);   <<< Pos - Start truncated
  }
  if (Err)
    *Err = createStringError(errc::illegal_byte_sequence,
                             "no null terminated string at offset 0x%" PRIx64,
                             Start);
  return StringRef();
}
```
There are just about 26,000 instances of `StringRef` in the repository (not counting tests) so I don't think reviewing them all is practical, but from what I've seen so far it doesn't look like a 32-bit LLVM can be expected to handle 64-bit code.

**Comment by efriedma-quic** [comment]
I would guess that in most cases, things appear to work because either values larger than 4GB are rare, or a value larger than 4GB would cause the compiler to run out of memory.  But there's probably a number of real issues.

I think it would be reasonable to change the relevant StringRef/ArrayRef APIs to take a 64-bit integer or have 64-bit overloads.

**Comment by msebor** [comment]
For reference (and for the test suite once this is fixed) the following shows invalid `strncmp` folding due to the `StringRef::substr()` truncation in an ILP32 compiler.  The expected result is for the `strncmp` call to be folded to a negative value.  That's also what a 64-bit host compiler does.
```
$ cat a.c && clang -O1 -S -Wall -emit-llvm -Xclang -disable-llvm-passes -o/dev/stdout a.c | /build/llvm-gcc-ilp32/bin/opt -S -instcombine
extern int strncmp (const char*, const char*, __SIZE_TYPE__);

const char a[] = "123", b[] = "124";
int f (void)
{
  return strncmp (a, b, ~0U + (__SIZE_TYPE__)1);
}

; ModuleID = '<stdin>'
source_filename = "a.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-redhat-linux-gnu"

@a = dso_local constant [4 x i8] c"123\00", align 1
@b = dso_local constant [4 x i8] c"124\00", align 1

; Function Attrs: nounwind uwtable
define dso_local i32 @f() #0 {
  ret i32 0
}

declare dso_local i32 @strncmp(i8*, i8*, i64) #1

attributes #0 = { nounwind uwtable "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { "frame-pointer"="none" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"uwtable", i32 1}
!2 = !{!"clang version 13.0.0 (Fedora 13.0.0-3.fc35)"}
```

**Comment by msebor** [comment]
> I would guess that in most cases, things appear to work because either values larger than 4GB are rare, or a value larger than 4GB would cause the compiler to run out of memory. But there's probably a number of real issues.
> 
> I think it would be reasonable to change the relevant StringRef/ArrayRef APIs to take a 64-bit integer or have 64-bit overloads.

To represent and manipulate strings longer than 4GB in a 32-bit LLVM, all of `StringRef` needs to change to use `uint64_t` instead of `size_t`, including the internal data members, and including `StringRef::npos`.  I made that change but the fallout from it is overwhelming.  Not just the number of errors due to ambiguous calls to functions like `std::min(T, T)` with different argument types, and the number of warnings due to assignments from `StringRef::npos` to `size_t` variables, but also, and most concerning, semantic changes in code that uses `std::string::npos` (260 instances) interchangeably with `StringRef::npos` (510 instances).  The `npos` problem, for instance, causes a build failure early on due the `llvm-tblgen` error:
```
llvm-tblgen: for the --help option: does not allow a value! 'help' specified.
```
Besides `StringRef` and related classes being pervasive, the code base is full of uses of `size_t` (I count over 22k instances of the type, not including tests, or those that are deduced to be `size_t` from their context like in `auto` declarations or in templates).  There are also a number of other classes that wrap `StringRef` and whose APIs interface with it.  Those APIs sometimes take `size_t` arguments that they pass to `StringRef` members.  An example is `CommandLineParser`and related classes.  It's difficult to tell which of those APIs might need to change to `uint64_t` as well and which ones are safe to leave unchanged.  Regardless of whether they change, all of their implementations need to be reviewed for the uses of `npos` to make sure it matches that of either `str::string` or `StringRef`.  With all that, this sort of a conversion doesn't seem feasible to me.

Unless someone has some bright ideas I'll go ahead and implement a fix targeted just at the `StringRef::substr()` problem that gave rise to this issue and also one for the `strncmp` bug above.

**Comment by nikic** [comment]
> To represent and manipulate strings longer than 4GB in a 32-bit LLVM, all of `StringRef` needs to change to use `uint64_t` instead of `size_t`, including the internal data members, and including `StringRef::npos`.

That doesn't really make sense to me. There is no way to store a string larger than 4GB on a 32-bit system anyway, so there should be no need for a uint64_t size member. This is *unlike* the originally reported issue, where the substr argument is just an upper-bound and does not actually require an allocation of that size -- we simply need to ignore substrs larger than SIZE_MAX there.

Though in any case, I don't think there is a lot of practical value in investing into edge-cases of the "cross-compiling from a 32-bit host to a 64-bit target" use-case.

**Comment by msebor** [comment]
Both 32-bit Clang and `opt` accept declarations of arrays larger than 4GB (e.g., `extern int a[4294967297LU];`).  I mentioned in PR  #54766 that initialized constants bigger than `UINT_MAX` bytes are silently miscompiled so there seems to be no way to create a C test case with one.  I also haven't been able to create a LLVM-only test case that creates one (it insists on providing initializers for all elements).  If not being to initialize such large constants was a deliberate limitation rather than a bug I'd expect Clang to give an error.  I took the comment in the review that gave rise to this bug and those in PR  #54766 to mean that those are supported use cases, and the comment here to suggest that the `StringRef` class should be changed to make it possible to eventually represent such large strings.  (Initializers for large constants such as in `const int a[SIZE_MAX + 2LU] = { [SIZE_MAX + 1] = 1 };` could be handled if they were represented as sparse arrays.)

But if there is consensus that initializing very large constants using a 32-bit Clang is in fact unsupported then (as I said) I'll just fix the two test cases that don't depend on that.

**Comment by msebor** [comment]
The two problems should be fixed by commit [ffed0cfc](https://github.com/llvm/llvm-project/commit/ffed0cfcdbee1b692fa19f9d97302cdf4423e5ee).
